apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1099,34 +1119,105 @@ private async Task ProcessPullMessagesAsync(List<ReceivedMessage> msgs, HashSet<
                 }
             }
 
-            private void HandleExtendLease(HashSet<string> msgIds)
+            private class LeaseCancellation : IDisposable
+            {
+                public LeaseCancellation(CancellationTokenSource softStopCts) =>
+                    _cts = CancellationTokenSource.CreateLinkedTokenSource(softStopCts.Token);
+
+                private readonly object _lock = new object();
+                private CancellationTokenSource _cts;
+
+                public CancellationToken Token
+                {
+                    get
+                    {
+                        lock (_lock)
+                        {
+                            return _cts?.Token ?? CancellationToken.None;
+                        }
+                    }
+                }
+
+                public void Dispose()
+                {
+                    lock (_lock)
+                    {
+                        _cts.Dispose();
+                        _cts = null;
+                    }
+                }
+
+                public bool IsDisposed
+                {
+                    get
+                    {
+                        lock (_lock)
+                        {
+                            return _cts == null;
+                        }
+                    }
+                }
+
+                public void Cancel()
+                {
+                    CancellationTokenSource cts2;
+                    lock (_lock)
+                    {
+                        cts2 = _cts;
+                    }
+                    cts2?.Cancel();
+                }
+            }
+
+            private void HandleExtendLease(HashSet<string> msgIds, LeaseCancellation cancellation)
             {
                 if (_softStopCts.IsCancellationRequested)
                 {
                     // No further lease extensions once stop is requested.
                     return;
                 }
-                bool anyMsgIds;
-                lock (msgIds)
+                if (cancellation == null)
                 {
-                    anyMsgIds = msgIds.Count > 0;
-                    if (anyMsgIds)
+                    // Create a task to cancel lease-extension once max-lease-extension time has been reached.
+                    cancellation = new LeaseCancellation(_softStopCts);
+                    Add(_scheduler.Delay(_maxExtensionDuration, cancellation.Token), Next(false, () =>
                     {
-                        lock (_lock)
+                        cancellation.Dispose();
+                        lock (msgIds)
                         {
-                            _extendQueue.Enqueue(msgIds.Select(x => new TimedId(_extendThrottleHigh + 1, x)));
+                            msgIds.Clear();
                         }
-                    }
+                    }));
```

Comment:
Should the method return at this point? Presumably it's just about feasible for _maxExtensionDuration to be one tick, so it could be disposed by the end of the block.